dlt/dlt_broker.py

```diff
@@ -82,17 +83,33 @@ def __init__(
             self.filter_ack_queue = None
             self.filter_ack_msg_handler = None
 
-        kwargs["ip_address"] = ip_address
-        kwargs["port"] = port
-        kwargs["timeout"] = kwargs.get("timeout", DLT_CLIENT_TIMEOUT)
-        self.msg_handler = DLTMessageHandler(
-            self.filter_queue,
-            self.message_queue,
-            self.mp_stop_flag,
-            kwargs,
-            dlt_time_value=self._dlt_time_value,
-            filter_ack_queue=self.filter_ack_queue,
-        )
+        self._from_dlt_client = True if ip_address else False
+
+        if self._from_dlt_client:
+            # If ip_address is given, then messages are retrieved from dlt client at run-time
+            kwargs["ip_address"] = ip_address
+            kwargs["port"] = port
+            kwargs["timeout"] = kwargs.get("timeout", DLT_CLIENT_TIMEOUT)
+            self.msg_handler = DLTMessageHandler(
+                self.filter_queue,
+                self.message_queue,
+                self.mp_stop_flag,
+                kwargs,
+                dlt_time_value=self._dlt_time_value,
+                filter_ack_queue=self.filter_ack_queue,
+            )
+        else:
+            # If not ip_address is given, then messages are retrieved from the given filename
+            # The logs are written to the given filename from another process
+            self.msg_handler = DLTFileSpinner(
+                self.filter_queue,
+                self.message_queue,
+                self.mp_stop_flag,
+                kwargs.get("filename"),
```

Comment:
I would promote filename as a named, optional parameter into the signature of the init function and update docstring there to describe what happens when it is passed and not the ip_address

Reply:
it could even better for the 2nd step in my plan:

```
class DLTBroker:
    def __init__(
       self,
       message_handler,
       enable_dlt_time=False,
       enable_filter_set_ack=False,
       filter_set_ack_timeout=2.0,
       ignore_filter_set_ack_timeout=False,
    ):
    self.msg_handler = message_handler

# Then for DLTConnector
message_handler = DLTMessageHandler(ip, port, timeout)
broker = DLTBroker(message_handler, ...)

# Then for DLTIOCConnector
message_handler = DLTFileSpinner(dlt_file)
broker = DLTBroker(message_handler, ...)
```

Sure, this will happen by touching DLTConnector part. So, I would leave it to the 2nd step, not now

Reply:
No need to wait for 2nd step

```python
class DLTBroker:
    @classmethod
    def create_from_file(cis, ...) -> "DLTBroker"
        message_handler = DLTFileSpinner(...)
        return cls(..., message_handler=message_handler)

    @classmethod
    def create_from_ip(cis, ...) -> "DLTBroker":
        message_handler = DLTMessageHandler(...)
       return cls(..., message_handler=message_handler)

    def __init__(
       self,
       enable_dlt_time=False,
       enable_filter_set_ack=False,
       filter_set_ack_timeout=2.0,
       ignore_filter_set_ack_timeout=False,
       message_handler: Optional[DLTMessageHandler] = None,
    ):
        self.message_handler = message_handler or ... # original code
```

Reply:
message_handler: Optional[DLTMessageHandler] could not be achieved as long as DLTFileSpinner has nothing to do with DLTMessageHandler at this moment, right?

Yeap, we could also have it defined without any type enforcement, but then the API is still not perfect until the 2nd ticket.

Reply:
so please read the type annotation slide. you can use `Union`

Reply:
I will make comments on the below discussion to consolidate both of these threads.

Reply:
After I take a deep look, I know that it's hard to pass an external DLTMessageHandler from outside. I don't like to init an object from different files. I purpose my solution. FYI

First, add initial functions in `dlt_broker_handlers.py`

```python
def _create_dlt_message_handler(
    ip_address,
    port,
    timeout,
    filter_queue,
    message_queue,
    mp_stop_flag,
    dlt_time_value,
    filter_ack_queue,
    **kwargs,
):
    kwargs["ip_address"] = ip_address
    kwargs["port"] = port
    kwargs["timeout"] = kwargs.get("timeout", DLT_CLIENT_TIMEOUT)

    return DLTMessageHandler(
        filter_queue,
        message_queue,
        mp_stop_flag,
        kwargs,
        dlt_time_value=dlt_time_value,
        filter_ack_queue=filter_ack_queue,
    )


def _create_dlt_file_spinner(filename, filter_queue, message_queue, mp_stop_flag, dlt_time_value, filter_ack_queue):
    return DLTFileSpinner(
        filter_queue,
        message_queue,
        mp_stop_flag,
        kwargs.get("filename"),
        dlt_time_value=dlt_time_value,
        filter_ack_queue=filter_ack_queue,
    )


def create_dlt_message_handler(
    ip_address,
    filename,
    filter_queue,
    message_queue,
    mp_stop_flag,
    dlt_time_value,
    filter_ack_queue,
    **kwargs,
) -> Union[DLTMessageHandler | DLTFileSpinner]:
    if ip_address:
        # If ip_address is given, then messages are retrieved from dlt client at run-time
        return _create_dlt_message_handler(
            ip_address,
            filter_queue,
            message_queue,
            mp_stop_flag,
            dlt_time_value,
            filter_ack_queue,
            **kwargs,
        )
    elif filename:
        # If not ip_address is given, then messages are retrieved from the given filename
        # The logs are written to the given filename from another process
        return _create_dlt_file_spinner(
            kwargs["filename"],
            filter_queue,
            message_queue,
            mp_stop_flag,
            dlt_time_value,
            filter_ack_queue,
            **kwargs,
        )

    raise NotImplementedError()
```

Second, change the `self.msg_handler` init function

```python
self.msg_handler = create_dlt_message_handler(
            ip_address,
            kwargs.get("filename"),
            self.filter_queue,
            self.message_queue,
            self.mp_stop_flag,
            self._dlt_time_value,
            self.filter_ack_queue,
            **kwargs,
        )
```